Common/lib/CANInterface/CANInterface.cpp

```diff
@@ -1,36 +1,44 @@
 #include "CANInterface.h"
 #include "Printing.h"
+#include "mbed.h"
+#include "rtos.h"
 
 #define CAN_PERIOD  10ms
 
-CANInterface::CANInterface(CAN &c, CANParser &cp, Ticker &tx, DigitalOut *stby, std::chrono::microseconds txPeriod) : can(c), canParser(cp), txTicker(tx)
-{
-    txTicker.attach(callback(this, &CANInterface::txHandler), (txPeriod==0s ? CAN_PERIOD : txPeriod));
-    can.attach(callback(this, &CANInterface::rxHandler), CAN::RxIrq);
-    if(stby)
-    {
+CANInterface::CANInterface(CAN &c, CANParser &cp, Thread &tx_thrd, Thread &rx_thrd, DigitalOut *stby, int txPeriod) : can(c), canParser(cp), tx_thread(rx_thrd), rx_thread(tx_thrd)
+{    
+    if(stby)    // should these be checked in the rx/tx threads? once we initialize this we 
+    {           // won't have control over it anymore 
         standby = stby;
         *standby = 0;   // active low
     }
 }
 
+void CANInterface::startCANTransmission(void)
+{
+    tx_thread.start(callback(this, &CANInterface::txHandler));
+    rx_thread.start(callback(this, &CANInterface::rxHandler));
+}
+
 // WARNING: This method will be called in an ISR context
-void CANInterface::rxHandler()
+void CANInterface::rxHandler(void)
 {
     CANMessage receivedCANMessage;
     while (can.read(receivedCANMessage))
     {
         canParser.parse(receivedCANMessage);
     }
+    ThisThread::sleep_for(CAN_PERIOD);
```

Comment:
There should be no waiting in the RX thread, so remove this line. This is because it will only cause longer delays for polling reads, which is undesirable. Also, our manual polling implementation is fine for now, but we really should look into setting up a CAN RX triggered interrupt (what we had for RX before) but which triggers an Mbed `Event`. This also uses some level of polling, but may be slightly faster than our implementation

Reply:
Sounds good. I will add an issue to move to a CAN RX triggered interrupt. 

On the topic of waiting in the tx_handler --> do we want to actually add a `sleep` hardcoded into the `tx_handler()`, or assume that the module will take care of CAN transmission periodicity on its own? If the main loop of the code module is running a 1khz task or something like that which is sending out CAN messages each tick of the statemachine, wouldn't we want to remove any sleep from the transmit thread and just let it handle transmitting messages as soon as the module cues them?

Reply:
And another thought --> do we think it makes sense to have the module set some sort of flag that the `CANInterface` can check when the `tx_handler` is called? This way, the `tx_handler` doesn't have to do any work unless there's actually a message to transmit. Might not be very efficient or effective in practice, but wanted to bounce the idea off.

Reply:
I think keeping the `sleep` in the `tx_handler()` is a good idea, as it keeps the code more structured, and the main loop can set the period for that sleep by passing in a period value to the constructor anyway.

Reply:
As for the flag, we can implement the skipping a message in the `CANParser` implementation by each board, by just simply returning a queue with fewer elements or no elements.